tools/pflua-optimize

```diff
@@ -18,26 +21,68 @@ it is read from standard input.
    print(content);
 end
 
--- Print help
-local input
+local function run(input, options)
+   local expr = assert(loadstring('return '..input:read('*a')))()
+
+   print('Optimizing:')
+   utils.pp(expr)
+   local optimized_expr = optimize.optimize(expr)
+   print('Result:')
+   utils.pp(optimized_expr)
+
+   if (options.pktfile)
+   then
+      local unopt_pred, opt_pred, packets, packet, P, len, unopt_res, opt_res
+      unopt_pred = pfcompile.compile_ast(expr, "Unopt")
+      opt_pred = pfcompile.compile_ast(optimized_expr, "Opt")
+      packets = savefile.load_packets(options.pktfile)
+      packet = packets[options.pktnum]
+      P, len = packet.packet, packet.len
 
-if #arg == 0 then
-   input = io.stdin
-elseif #arg == 1 then
-   if arg[1] == '--help' or arg[1] == '-h' then
+      unopt_res = unopt_pred(P, len)
+      opt_res = opt_pred(P, len)
+      if unopt_res ~= opt_res
+      then
+         print("Packet results did not match!")
+         print("Unoptimized: %s, optimized: %s"):format(unopt_res, opt_res)
+         exit(1)
+      else
+         print("Match status the same before and after optimization.")
+      end
+   end
+end
+
+local function initialize_from_command_line(args)
+   local input
+   local options = {}
+   while #args >= 1 and args[1]:match("^%-%-") do
+      local param_arg = table.remove(args, 1)
+      if param_arg == '--match-packet'
+      then
+         options.pktnum = tonumber(table.remove(args, 1))
+         options.pktfile = table.remove(args, 1)
+         print(options.pktnum, options.pktfile)
+      else error("Unknown argument: " .. arg) end
+   end
+
+   if #args == 0 then
+      input = io.stdin
+   elseif #args == 1 then
+      if args[1] == '--help' or args[1] == '-h' then
+         usage()
+         os.exit(0)
+      end
+      input = assert(io.open(args[1]))
+   else
       usage()
-      os.exit(0)
+      os.exit(1)
    end
-   input = assert(io.open(arg[1]))
-else
-   usage()
-   os.exit(1)
+   return input, options
 end
 
-local expr = assert(loadstring('return '..input:read('*a')))()
+local function main(...)
+   local args = { ... }
+   run(initialize_from_command_line(args))
```

Comment:
neat, i didn't know that you could return multiple values from initialize() and have them passed to run() as multiple arguments!

I think I would rename `initialize_from_command_line` to `parse_command_line` or so -- initialize connotes a stateful transformation, whereas here the function has no side effects -- you have to use the result values.